### PR TITLE
STORM-2867: Add consumer lag metrics to KafkaSpout

### PR DESCRIPTION
**Pull request by omkreddy**

### CODE REVIEW

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/KafkaUtils.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+import org.apache.storm.metric.api.IMetric;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+
+public class KafkaUtils {
```
Comment (hmcl) [inline external/storm-kafka-client/src/main/java/org/apache/storm/kafka/KafkaUtils.java]: Why the need of the KafkaUtils wrapper class? Why not just have the KafkaOffsetMetric class ?

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/KafkaUtils.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+                    Set<TopicPartition> topicPartitions = offsetManagers.keySet();
+
+                    Map<TopicPartition, Long> beginningOffsets= kafkaConsumer.beginningOffsets(topicPartitions);
+                    Map<TopicPartition, Long> endOffsets= kafkaConsumer.endOffsets(topicPartitions);
+
+                    for (Map.Entry<TopicPartition, OffsetManager> entry : offsetManagers.entrySet()) {
+                        TopicPartition topicPartition = entry.getKey();
+                        OffsetManager offsetManager = entry.getValue();
+
+                        long latestTimeOffset = beginningOffsets.get(topicPartition);
```
Comment (HeartSaVioR) [inline external/storm-kafka-client/src/main/java/org/apache/storm/kafka/KafkaUtils.java]: Looks like twos are swapped: beginning -> earliest, end -> latest

https://kafka.apache.org/0101/javadoc/org/apache/kafka/clients/consumer/KafkaConsumer.html#beginningOffsets(java.util.Collection)

https://kafka.apache.org/0101/javadoc/org/apache/kafka/clients/consumer/KafkaConsumer.html#endOffsets(java.util.Collection)

File: examples/storm-kafka-client-examples/pom.xml [hunk 3]
```diff
@@ -73,12 +73,6 @@
             <version>${project.version}</version>
             <scope>${provided.scope}</scope>
         </dependency>
-        <dependency>
```
Comment (HeartSaVioR) [inline examples/storm-kafka-client-examples/pom.xml]: Do you run example topology after this change? If then I'm OK. Just to verify the change doesn't hurt anything.

File: external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/KafkaSpoutConfigTest.java [hunk 4]
```diff
@@ -49,6 +49,7 @@ public void testBasic() {
         expected.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
         expected.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
         assertEquals(expected, conf.getKafkaProps());
+        assertEquals(KafkaSpoutConfig.DEFAULT_METRICS_TIME_BUCKET_SIZE_SECONDS, conf.getMetricsTimeBucketSizeInSecs());
```
Comment (satishd) [inline external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/KafkaSpoutConfigTest.java]: This may be in plan to add in next commits. You may want to add a test to see whether the configured property is also taken into account.

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/KafkaUtils.java (hunk 1 above)
Comment (satishd) [reply external/storm-kafka-client/src/main/java/org/apache/storm/kafka/KafkaUtils.java]: May be a copy/paste issue, taken from https://github.com/apache/storm/blob/master/external/storm-kafka/src/jvm/org/apache/storm/kafka/KafkaUtils.java

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutConfig.java [hunk 5]
```diff
@@ -182,6 +186,7 @@ public KafkaSpoutConfig(Builder<K, V> builder) {
         private boolean emitNullTuples = false;
         private ProcessingGuarantee processingGuarantee = DEFAULT_PROCESSING_GUARANTEE;
         private boolean tupleTrackingEnforced = false;
+        public int metricsTimeBucketSizeInSecs = DEFAULT_METRICS_TIME_BUCKET_SIZE_SECONDS;
```
Comment (satishd) [inline external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutConfig.java]: access of this field should be private instead of public.

File: examples/storm-kafka-client-examples/pom.xml (hunk 3 above)
Comment (srdo) [inline examples/storm-kafka-client-examples/pom.xml]: Why was this removed?

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/KafkaUtils.java (hunk 1 above)
Comment (srdo) [reply external/storm-kafka-client/src/main/java/org/apache/storm/kafka/KafkaUtils.java]: It would be better to make a package for this (e.g. org.apache.storm.kafka.spout.metrics?), instead of using the KafkaUtils class for namespacing.

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/KafkaUtils.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+import java.util.Set;
+import java.util.TreeMap;
+
+public class KafkaUtils {
+
+    private static final Logger LOG = LoggerFactory.getLogger(KafkaUtils.class);
+
+    public static class KafkaOffsetMetric implements IMetric {
+        Map<TopicPartition, OffsetManager> offsetManagers;
+        KafkaConsumer kafkaConsumer;
```
Comment (srdo) [inline external/storm-kafka-client/src/main/java/org/apache/storm/kafka/KafkaUtils.java]: This will break. The spout replaces the consumer when it is deactivated, so cycling topology activation will break this metric. I would either use a KafkaConsumer supplier here, or add a setter to this metric.

Also why not private?

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/KafkaUtils.java [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        private class TopicMetrics {
+            long totalSpoutLag = 0;
+            long totalEarliestTimeOffset = 0;
+            long totalLatestTimeOffset = 0;
+            long totalLatestEmittedOffset = 0;
+            long totalLatestCompletedOffset = 0;
+        }
+
+        @Override
+        public Object getValueAndReset() {
```
Comment (srdo) [inline external/storm-kafka-client/src/main/java/org/apache/storm/kafka/KafkaUtils.java]: Remind me if this method is called by a different thread from the executor thread that runs nextTuple and the other spout methods? If so we need to add some synchronization on the offset managers. If not I don't think we need a ConcurrentHashMap in the spout.

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/internal/OffsetManager.java [hunk 8]
```diff
@@ -57,7 +58,8 @@ public void addToAckMsgs(KafkaSpoutMessageId msgId) {          // O(Log N)
     }
 
     public void addToEmitMsgs(long offset) {
-        this.emittedOffsets.add(offset);                  // O(Log N)
+        this.emittedOffsets.add(offset);  // O(Log N)
+        this.latestEmittedOffset = offset;
```
Comment (srdo) [inline external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/internal/OffsetManager.java]: I don't think this works. This will just set the latest emitted offset to the offset that was most recently emitted. I think you want the highest offset that has been emitted.

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/KafkaUtils.java [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+                        ret.put(topic + "/" + "totalLatestTimeOffset", topicMetrics.totalLatestTimeOffset);
+                        ret.put(topic + "/" + "totalLatestEmittedOffset", topicMetrics.totalLatestEmittedOffset);
+                        ret.put(topic + "/" + "totalLatestCompletedOffset", topicMetrics.totalLatestCompletedOffset);
+                    }
+
+                    return ret;
+                } else {
+                    LOG.info("Metrics Tick: Not enough data to calculate spout lag.");
+                }
+            } catch (Throwable t) {
```
Comment (srdo) [inline external/storm-kafka-client/src/main/java/org/apache/storm/kafka/KafkaUtils.java]: Why are we catching all exceptions here?

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java [hunk 10]
```diff
@@ -125,12 +129,22 @@ public void open(Map conf, TopologyContext context, SpoutOutputCollector collect
...
         refreshSubscriptionTimer = new Timer(TIMER_DELAY_MS, kafkaSpoutConfig.getPartitionRefreshPeriodMs(), TimeUnit.MILLISECONDS);
 
-        offsetManagers = new HashMap<>();
+        offsetManagers = new ConcurrentHashMap<>();
         emitted = new HashSet<>();
         waitingToEmit = Collections.emptyListIterator();
 
         tupleListener.open(conf, context);
 
+        context.registerMetric("kafkaOffset", new IMetric() {
```
Comment (srdo) [inline external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java]: Why do we need the wrapper metric?

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java [hunk 11]
```diff
@@ -553,9 +567,11 @@ public void fail(Object messageId) {
     public void activate() {
         try {
             subscribeKafkaConsumer();
+            metricsConsumer = new KafkaConsumer<>(kafkaSpoutConfig.getKafkaProps());
```
Comment (srdo) [inline external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java]: Why do we need a separate consumer for the metric? Also if we really need two consumers, you should create this via the factory. Otherwise you prevent us from stubbing/spying on this consumer.

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java [hunk 12]
```diff
@@ -568,6 +584,8 @@ private void subscribeKafkaConsumer() {
     public void deactivate() {
         try {
             shutdown();
+            metricsConsumer.close();
```
Comment (srdo) [inline external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java]: I think this belongs in shutdown.

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutConfig.java [hunk 13]
```diff
@@ -667,6 +672,16 @@ private void setNonNullSerDeKafkaProp(SerializableDeserializer<K> keyDes, Class<
             return this;
         }
 
+        /**
+         * The time period that metrics data in bucketed into.
+         * @param metricsTimeBucketSizeInSecs
+         * @return
```
Comment (srdo) [inline external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutConfig.java]: Nit: Don't add empty `@return`, just leave it off.

File: pom.xml [hunk 14]
```diff
@@ -278,7 +278,7 @@
         <storm.kafka.artifact.id>kafka_2.10</storm.kafka.artifact.id>
 
         <!-- kafka version used by new storm-kafka-client spout code -->
-        <storm.kafka.client.version>0.10.0.0</storm.kafka.client.version>
+        <storm.kafka.client.version>0.11.0.2</storm.kafka.client.version>
```
[the remark this answers is not in the document]
Comment (srdo) [reply pom.xml]: We need to upgrade to at least 0.10.1.0, since that's where the `beginningOffsets` and `endOffsets` methods were added to the KafkaConsumer. I think upgrading the consumer and adding this patch would probably be a breaking change, because the spout won't work with 0.10.0.x client libraries/brokers anymore.

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/KafkaUtils.java (hunk 1 above)
Comment (omkreddy) [reply external/storm-kafka-client/src/main/java/org/apache/storm/kafka/KafkaUtils.java]: Removed KafkaUtils wrapper. moved KafkaOffsetMetric to org.apache.storm.kafka.spout.metrics package

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/KafkaUtils.java (hunk 2 above)
Comment (omkreddy) [reply external/storm-kafka-client/src/main/java/org/apache/storm/kafka/KafkaUtils.java]: Updated the code

File: pom.xml (hunk 14 above)
[the remark this answers is not in the document]
Comment (omkreddy) [reply pom.xml]: yes, we need at-least 0.10.1.0 kafka clients and its a breaking change. We check for any other options

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/KafkaUtils.java [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+                        long spoutLag = latestTimeOffset - latestCompletedOffset;
+
+                        String metricPath = topicPartition.topic()  + "/partition_" + topicPartition.partition();
+                        ret.put(metricPath + "/" + "spoutLag", spoutLag);
+                        ret.put(metricPath + "/" + "earliestTimeOffset", earliestTimeOffset);
+                        ret.put(metricPath + "/" + "latestTimeOffset", latestTimeOffset);
+                        ret.put(metricPath + "/" + "latestEmittedOffset", latestEmittedOffset);
+                        ret.put(metricPath + "/" + "latestCompletedOffset", latestCompletedOffset);
+
+                        if (!topicMetricsMap.containsKey(topicPartition.topic())) {
```
[the remark this answers is not in the document]
Comment (omkreddy) [reply external/storm-kafka-client/src/main/java/org/apache/storm/kafka/KafkaUtils.java]: Done

File: external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/KafkaSpoutConfigTest.java (hunk 4 above)
Comment (omkreddy) [reply external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/KafkaSpoutConfigTest.java]: added a small testcase

File: examples/storm-kafka-client-examples/pom.xml (hunk 3 above)
Comment (omkreddy) [reply examples/storm-kafka-client-examples/pom.xml]: I think we don't need kafka server jar to run  examples. Ran the tests without this dependency. But i reverted the change, since it is not related to this PR.

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/KafkaUtils.java (hunk 9 above)
Comment (omkreddy) [reply external/storm-kafka-client/src/main/java/org/apache/storm/kafka/KafkaUtils.java]: Removed Throwable from catch clause.

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java (hunk 10 above)
Comment (omkreddy) [reply external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java]: Removed the wrapper metric.

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/KafkaUtils.java (hunk 7 above)
Comment (omkreddy) [reply external/storm-kafka-client/src/main/java/org/apache/storm/kafka/KafkaUtils.java]: Thats good point. Will check  if it is called from same thread or different thread, based on that I will remove ConcurrentHashMap usage. We also don't need separate consumer for the metric, it it is called from the same thread. Will check and update the PR.

Comment (arunmahadevan) [reply external/storm-kafka-client/src/main/java/org/apache/storm/kafka/KafkaUtils.java]: @omkreddy , It seems both are invoked from the spout executor thread. The getValueAndReset is invoked from the tuple-action-fn which is invoked from [here](https://github.com/apache/storm/blob/1.x-branch/storm-core/src/clj/org/apache/storm/daemon/executor.clj#L627) and the nextTuple [here](https://github.com/apache/storm/blob/1.x-branch/storm-core/src/clj/org/apache/storm/daemon/executor.clj#L647)

File: pom.xml (hunk 14 above)
[the remark this answers is not in the document]
Comment (HeartSaVioR) [reply pom.xml]: Actually we are "effectively" dropping support of Kafka 0.9.x while moving from Storm 1.0.x to 1.1.0, because critical bugs in storm-kafka-client are not ported back to 1.0.x version line. I think we should try to avoid doing it again, but not sure for the cases like this. (bump the version for utilizing new feature)

Btw, if Kafka doesn't always guarantee forward compatibility, we may need to define support versions of Kafka (and others, if we could maintain) and maintain that.

**Comment by srdo** [comment]
Regarding the Kafka upgrade in this:

I think we can keep support for 0.10.0.x by adding a check to KafkaSpout.open that checks if the two new consumer methods are present via reflection, and disables the new metric if they aren't. Ideally we'd be able to use this type of approach any time we add a feature that uses a new method on the KafkaConsumer.

It would be fine by me if we only do this in 1.x to retain backward compatibility, and just leave out the check in master. Maybe we add a jira issue to 2.0.0 that says we're requiring at least 0.10.1.0, so it's visible to users upgrading to 2.0.0. 

If we really need to support old Kafka versions indefinitely, we can hopefully just keep the reflection checks, but I'd rather use Storm major versions as an opportunity to clean up the code and require newer Kafka versions.

I believe Kafka guarantees interface stability for KafkaConsumer as of 0.10.0.0. Since users can specify which kafka-clients version they want when they define their topologies, we should not need to maintain different versions of storm-kafka-client for different Kafka versions.

**Comment by HeartSaVioR** [comment]
+1 on @srdo 's approach. Btw, Flink has Kafka connector modules for each major version, so if necessary, we could try out similar approach. If we want to separate 0.10.x and 0.11.x+, it might be leveraging support of exactly-once semantics in Kafka 0.11.0+.

**Comment by srdo** [comment]
Sure, sounds good to me. If we can get away with a little bit of reflection I'd rather try that first, but putting out a storm-kafka-client-0.10.1, storm-kafka-client-1.0.0 etc. would be good too, as long as we have a storm-kafka-client-base project for most of the code. I just don't want to get into a situation where we have to maintain 2-3 Storm branch versions * 2-3 storm-kafka-client variants per branch.

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/metrics/KafkaOffsetMetric.java [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+public class KafkaOffsetMetric implements IMetric {
+
+    private static final Logger LOG = LoggerFactory.getLogger(KafkaOffsetMetric.class);
+    private Map<TopicPartition, OffsetManager> offsetManagers;
```
Comment (srdo) [inline external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/metrics/KafkaOffsetMetric.java]: nit: It might be better to use a getter or Supplier to access the field in KafkaSpout, rather than copying the Map reference in here. If someone does e.g. `offsetManagers = new HashMap<>()` in the spout, this reference ends up pointing to the wrong thing.

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/metrics/KafkaOffsetMetric.java [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+                    TopicMetrics topicMetrics = e.getValue();
+                    ret.put(topic + "/" + "totalSpoutLag", topicMetrics.totalSpoutLag);
+                    ret.put(topic + "/" + "totalEarliestTimeOffset", topicMetrics.totalEarliestTimeOffset);
+                    ret.put(topic + "/" + "totalLatestTimeOffset", topicMetrics.totalLatestTimeOffset);
+                    ret.put(topic + "/" + "totalLatestEmittedOffset", topicMetrics.totalLatestEmittedOffset);
+                    ret.put(topic + "/" + "totalLatestCompletedOffset", topicMetrics.totalLatestCompletedOffset);
+                }
+                LOG.debug("Metrics Tick: value : {}", ret);
+                return ret;
+            } else {
```
Comment (srdo) [inline external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/metrics/KafkaOffsetMetric.java]: Nit: Invert the if-else here to reduce indentation. It also allows you to avoid the "catch-all" null return in the bottom.

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/metrics/KafkaOffsetMetric.java [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+                LOG.info("Metrics Tick: Not enough data to calculate spout lag.");
+            }
+        } catch (Exception t) {
```
Comment (srdo) [inline external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/metrics/KafkaOffsetMetric.java]: Sorry, I wasn't being clear with my earlier comment: Why are we catching exceptions here at all?

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/metrics/KafkaOffsetMetric.java [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+import org.apache.storm.kafka.spout.internal.OffsetManager;
+import org.apache.storm.metric.api.IMetric;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+public class KafkaOffsetMetric implements IMetric {
```
Comment (srdo) [inline external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/metrics/KafkaOffsetMetric.java]: Could we add a test or two for this class? Maybe a test in KafkaSpoutSingleTopicTest that starts up a spout and checks that it can get metrics for the subscribed partition?

Comment (omkreddy) [reply external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/metrics/KafkaOffsetMetric.java]: Added a simple testcase

**Comment by omkreddy** [comment]
@srdo Thanks for the detailed review and suggestions. Updated the PR with review comments. Used refections approach to enable or disable metrics.  Pls take a look.

cc @HeartSaVioR  @satishd

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/metrics/KafkaOffsetMetric.java [hunk 20]
```diff
@@ -31,77 +32,75 @@
 public class KafkaOffsetMetric implements IMetric {
 
     private static final Logger LOG = LoggerFactory.getLogger(KafkaOffsetMetric.class);
-    private Map<TopicPartition, OffsetManager> offsetManagers;
-    private KafkaConsumer kafkaConsumer;
+    private Supplier<Map<TopicPartition, OffsetManager>> offsetManagerSupplier;
```
Comment (srdo) [inline external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/metrics/KafkaOffsetMetric.java]: Nit: These can be final now

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/metrics/KafkaOffsetMetric.java [hunk 21]
```diff
@@ -31,77 +32,75 @@
...
+            topicMetrics.totalLatestTimeOffset += latestTimeOffset;
+            topicMetrics.totalLatestEmittedOffset += latestEmittedOffset;
+            topicMetrics.totalLatestCompletedOffset += latestCompletedOffset;
+        }
+
+        for(Map.Entry<String, TopicMetrics> e : topicMetricsMap.entrySet()) {
+            String topic = e.getKey();
+            TopicMetrics topicMetrics = e.getValue();
+            result.put(topic + "/" + "totalSpoutLag", topicMetrics.totalSpoutLag);
+            result.put(topic + "/" + "totalEarliestTimeOffset", topicMetrics.totalEarliestTimeOffset);
```
Comment (srdo) [inline external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/metrics/KafkaOffsetMetric.java]: I'm not sure I understand a few of these metrics. What does e.g. totalEarliestTimeOffset tell the user? To me it looks like totalSpoutLag makes a lot of sense to have, but the other metric totals aren't very meaningful.

Just to give an example, if I get out a value of 10 for totalLatestTimeOffset, I don't understand what use that could have, because it's just telling me that if I add together the log end offsets of my assigned partitions, I get 10. It could be used to calculate how many messages are in all the partitions, but in that case it makes more sense to return e.g. `totalRecordsInAssignedPartitions = totalLatestTimeOffset - totalEarliestTimeOffset - 1` (disregarding log compaction).

File: external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/KafkaSpoutSingleTopicTest.java [hunk 22]
```diff
@@ -343,4 +344,48 @@ public void testSpoutMustRefreshPartitionsEvenIfNotPolling() throws Exception {
...
+        assertEquals(0, offsetMetric.get(SingleTopicKafkaSpoutConfiguration.TOPIC+"/totalEarliestTimeOffset").longValue());
+        // the offset of the last available message + 1.
+        assertEquals(10, offsetMetric.get(SingleTopicKafkaSpoutConfiguration.TOPIC+"/totalLatestTimeOffset").longValue());
+        assertEquals(0, offsetMetric.get(SingleTopicKafkaSpoutConfiguration.TOPIC+"/totalLatestEmittedOffset").longValue());
+        assertEquals(0, offsetMetric.get(SingleTopicKafkaSpoutConfiguration.TOPIC+"/totalLatestCompletedOffset").longValue());
+        //totalSpoutLag = totalLatestTimeOffset-totalLatestCompletedOffset
+        assertEquals(10, offsetMetric.get(SingleTopicKafkaSpoutConfiguration.TOPIC+"/totalSpoutLag").longValue());
+
+        //Emit and Ack all messages
+        for(int i = 0; i < messageCount; i++) {
```
Comment (srdo) [inline external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/KafkaSpoutSingleTopicTest.java]: I think you can use nextTuple_verifyEmitted_ack_resetCollector(int offset) from the superclass to do everything in this block.

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java [hunk 23]
```diff
@@ -739,4 +764,9 @@ public boolean shouldPoll() {
             return !this.pollablePartitions.isEmpty();
         }
     }
+
+    @VisibleForTesting
+    KafkaOffsetMetric getKafkaOffsetMetric() {
```
Comment (srdo) [inline external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java]: Optional: We could avoid this getter by using the same trick with a factory as we use for injecting the KafkaConsumer. If we add a KafkaOffsetMetricFactory to the spout constructor, you can create the metric in the test and pass it into the spout instead of letting the spout construct the metric.

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/metrics/KafkaOffsetMetric.java (hunk 21 above)
Comment (omkreddy) [reply external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/metrics/KafkaOffsetMetric.java]: @srdo  I agree. totalEarliestTimeOffset alone is not much useful. I just added it to make the metrics compatible with old spout (storm-kafka) metrics. 

adding totalRecordsInAssignedPartitions is good idea. I have included totalRecordsInAssignedPartitions to metrics. Can you explain why we need to decrease the count by 1?  I am not understood what do you mean by disregarding log compaction?

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java (hunk 23 above)
Comment (omkreddy) [reply external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java]: Agreed,  but I feel current approach is simpler. maybe we can evolve when we add more metrics.

Comment (srdo) [reply external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java]: Yes, makes sense. Thanks.

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/metrics/KafkaOffsetMetric.java (hunk 21 above)
Comment (srdo) [reply external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/metrics/KafkaOffsetMetric.java]: The reason I'm asking is because I'd rather only copy over the useful metrics from storm-kafka. It's easier to not add metrics we don't think are useful, than to remove them later.

Sorry, I made a mistake. We don't need to subtract 1. My bad.

Regarding why we disregard log compaction, if there's a partition with a log end offset (latestTimeOffset) of 10, and a beginningOffset of 0, we'd expect there to be 10 messages. When log compaction is enabled, some of the messages between 0 and 10 may have been replaced by tombstones, so they're not there anymore. The metric will still show 10 messages present, because we have no way to know if some of the messages between the beginningOffset and log end offset have been compacted away.

Great, thanks for taking the time to verify that the reflection code works.

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/metrics/KafkaOffsetMetric.java [hunk 24]
```diff
@@ -0,0 +1,118 @@
...
+            long spoutLag = latestTimeOffset - latestCompletedOffset;
+            long recordsInAssignedPartitions =  latestTimeOffset - earliestTimeOffset;
+
+            String metricPath = topicPartition.topic()  + "/partition_" + topicPartition.partition();
+            result.put(metricPath + "/" + "spoutLag", spoutLag);
+            result.put(metricPath + "/" + "earliestTimeOffset", earliestTimeOffset);
+            result.put(metricPath + "/" + "latestTimeOffset", latestTimeOffset);
+            result.put(metricPath + "/" + "latestEmittedOffset", latestEmittedOffset);
+            result.put(metricPath + "/" + "latestCompletedOffset", latestCompletedOffset);
+            result.put(metricPath + "/" + "recordsInAssignedPartitions", recordsInAssignedPartitions);
```
Comment (srdo) [inline external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/metrics/KafkaOffsetMetric.java]: Nit: Sorry, now that I look at it the name here is a little weird. Maybe `recordsInPartition`/`totalRecordsInPartitions`?

Comment (omkreddy) [reply external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/metrics/KafkaOffsetMetric.java]: Updated the PR.

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java (hunk 23 above)
Comment (hmcl) [reply external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java]: If we start adding a lot of these test methods we would be better off but creating a class in the tests packaged called KafkaSpoutTest that extends KafkaSpout and use that one in the tests. All of these methods should go in this class. WE don't want this class to be very bloated.

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/metrics/KafkaOffsetMetric.java [hunk 25]
```diff
@@ -0,0 +1,118 @@
...
+    }
+
+    @Override
+    public Object getValueAndReset() {
+
+        Map<TopicPartition, OffsetManager> offsetManagers = offsetManagerSupplier.get();
+        KafkaConsumer kafkaConsumer = consumerSupplier.get();
+
+        if (offsetManagers == null || offsetManagers.isEmpty() || kafkaConsumer == null) {
+            LOG.info("Metrics Tick: offsetManagers or kafkaConsumer is null.");
```
Comment (hmcl) [inline external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/metrics/KafkaOffsetMetric.java]: Should this be INFO level? Is this going to print this message periodically?

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/metrics/KafkaOffsetMetric.java [hunk 26]
```diff
@@ -0,0 +1,118 @@
...
+            return null;
+        }
+
+        Map<String,TopicMetrics> topicMetricsMap = new HashMap<>();
+        Set<TopicPartition> topicPartitions = offsetManagers.keySet();
+
+        Map<TopicPartition, Long> beginningOffsets= kafkaConsumer.beginningOffsets(topicPartitions);
+        Map<TopicPartition, Long> endOffsets= kafkaConsumer.endOffsets(topicPartitions);
+        Map<String, Long> result = new HashMap<>();
```
Comment (hmcl) [inline external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/metrics/KafkaOffsetMetric.java]: it would be useful to have a comment saying what is in this result map

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/metrics/KafkaOffsetMetric.java [hunk 27]
```diff
@@ -0,0 +1,118 @@
...
+import java.util.Map;
+import java.util.Set;
+
+public class KafkaOffsetMetric implements IMetric {
+
+    private static final Logger LOG = LoggerFactory.getLogger(KafkaOffsetMetric.class);
+    private final Supplier<Map<TopicPartition, OffsetManager>> offsetManagerSupplier;
+    private final Supplier<KafkaConsumer> consumerSupplier;
+
+    public KafkaOffsetMetric(Supplier offsetManagerSupplier, Supplier consumerSupplier) {
```
Comment (hmcl) [inline external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/metrics/KafkaOffsetMetric.java]: What's the reasoning behind passing Supplier rather than the actual object?

Comment (srdo) [reply external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/metrics/KafkaOffsetMetric.java]: The consumer is replaced when activating/deactivating the spout, so we can't just pass the object in here. The alternative is to have a setConsumer method on the metric that the spout then has to call. For the offset manager map, see https://github.com/apache/storm/pull/2480#discussion_r158958224.

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/metrics/KafkaOffsetMetric.java (hunk 25 above)
Comment (omkreddy) [reply external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/metrics/KafkaOffsetMetric.java]: This can be at info level. updated the PR.

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/metrics/KafkaOffsetMetric.java (hunk 26 above)
Comment (omkreddy) [reply external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/metrics/KafkaOffsetMetric.java]: added a comment and added java docs to the classs

**Comment by srdo** [comment]
@omkreddy This LGTM. Could you open a PR against master as well, we usually put features on that branch before merging to the other versions?

**Comment by omkreddy** [comment]
@srdo  Raised PR for master : https://github.com/apache/storm/pull/2495

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java [hunk 28]
```diff
@@ -144,10 +147,37 @@ public void open(Map conf, TopologyContext context, SpoutOutputCollector collect
         setCommitMetadata(context);
 
         tupleListener.open(conf, context);
+        if (canRegisterMetrics()) registerMetric();
```
Comment (srdo) [inline external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java]: Nit: Use braces.

Comment (omkreddy) [reply external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java]: Updated the PR.

**Comment by arunmahadevan** [comment]
+1
